job-runner/chain/traverser.go

```diff
@@ -146,18 +165,31 @@ func (t *traverser) Run() error {
 
 	var finalState byte
 	defer func() {
-		// Set final state of chain in repo. This will be very short-lived because
-		// next we'll finalize the request in the RM. Although short-lived, we set
-		// it in case there's problems finalizing with RM.
-		t.chain.SetState(finalState)
-		t.chainRepo.Set(t.chain)
-
-		// Mark the request as finished in the Request Manager.
-		if err := t.rmc.FinishRequest(t.chain.RequestId(), finalState); err != nil {
-			t.logger.Errorf("problem reporting status of the finished chain: %s", err)
-		} else {
-			t.chainRepo.Remove(t.chain.RequestId())
+		// Only tell RM the request was finished if job chain wasn't suspended or
+		// if it was suspended but completed successfully anyways.
+		if !t.suspended() || finalState == proto.STATE_COMPLETE {
+			// Set final state of chain in repo. This will be very short-lived
+			// because next we'll finalize the request in the RM. Although
+			// short-lived, we set it in case there's problems finalizing with RM.
+			t.chain.SetState(finalState)
+			t.chainRepo.Set(t.chain)
+
+			// Mark the request as finished in the Request Manager.
+			if err := t.rmc.FinishRequest(t.chain.RequestId(), finalState); err != nil {
+				t.logger.Errorf("problem reporting status of the finished chain: %s", err)
+			} else {
+				t.chainRepo.Remove(t.chain.RequestId())
+			}
 		}
+
+		// Job chain is done running - doneChan is watched by suspend().
+		close(t.doneChan)
+	}()
+
+	// Watch for shutdown signal and suspend job chain when received.
+	go func() {
+		<-t.shutdownChan
+		t.suspend()
```

Comment:
Same as a comment below: "shutdown" makes more sense to me because we're not suspending the traverser (we suspend job chains).  I.e. trying to keep `suspend*` a term related just to job chains, not to other things.